Sources/Build/MakefileSupport.swift

```diff
@@ -0,0 +1,553 @@
+        stream <<< "$(BUILD_DIR)/" <<< product.name
+        stream <<< "\n"
+
+        // Phony rule.
+        stream <<< "\n"
+        stream <<< ".PHONY: " <<< product.ruleName <<< "\n"
+        stream <<< product.ruleName <<< ": " <<< product.makeVarSub("EXE") <<< "\n"
+
+        // Compute the object that needs to be linked.
+        let objVarSubs = recursiveDependencies(of: product).map({ $0.makeVarSub("OBJS") }).joined(separator: " ")
+
+        // Main rule.
+        stream <<< "\n"
+        stream <<< product.makeVarSub("EXE") <<< ": " <<< objVarSubs <<< "\n"
+
+        stream <<< "\t@echo Linking " <<< product.name <<< "\n"
+
+        // This is only required for macOS to create the bundle structure.
+        if product.type == .test {
+            stream <<< "\t@mkdir -p $(dir \(product.makeVarSub("EXE")))" <<< "\n"
+        }
+
+        // Product specific args.
+        let productArgs: [String]
+        switch product.type {
+        case .executable:
+            productArgs = ["-emit-executable"]
+        case .test:
+            productArgs = ["$(TEST_LINKER_ARGS)"]
+        case .library:
+            productArgs = []
+        }
+
+        stream <<< """
+            \t@$(SWIFT_EXEC) -target $(TARGET) -sdk $(SDKROOT) \
+            $(EXTRA_SWIFT_FLAGS) -g -L $(BUILD_DIR) -o \(product.makeVarSub("EXE")) \
+            -module-name \(product.name) \(productArgs.joined(separator: " ")) \(objVarSubs)
+            """
+        stream <<< "\n"
+        stream <<< "# --------------------------------------------------------------- #" <<< "\n"
+    }
+
+    private func appendTargetDependencies(_ targets: [ResolvedTarget], to stream: OutputByteStream) {
+        stream <<< "\n"
+        stream <<< "# Dependencies" <<< "\n"
+        for target in targets {
+             switch target.underlyingTarget {
+             case is SwiftTarget:
+                appendTargetDependency(target, stream: stream)
+             case is ClangTarget:
+                 break
+             case is SystemLibraryTarget:
+                 break
+             default:
+                 fatalError("unhandled \(target.underlyingTarget)")
+             }
+        }
+    }
+
+    private func appendTargetDependency(_ target: ResolvedTarget, stream: OutputByteStream) {
+        // Add dependencies.
+        for dependency in target.dependencies {
+            switch dependency {
+            case .target(let dependencyTarget):
+                stream <<< target.makeVarSub("SWIFT_MODULE") <<< ": " <<< dependencyTarget.makeVarSub("SWIFT_MODULE") <<< "\n"
+            case .product:
+                fatalError("Unsupported")
+            }
+        }
+    }
+
+    private func appendSwiftRule(_ target: ResolvedTarget, linuxOnly: Bool, stream: OutputByteStream) {
+        stream <<< "\n"
+        stream <<< "# --------------------------------------------------------------- #" <<< "\n"
+        if linuxOnly {
+            stream <<< "ifeq ($(shell uname -s), Linux)\n"
```

Comment:
Probably better to factor out a PLATFORM make variable earlier, and use that consistently instead of uname